flink-core/src/main/java/org/apache/flink/core/memory/HybridMemorySegment.java

```diff
@@ -132,10 +132,14 @@
     @Override
     public void free() {
         super.free();
+        runCleanerAtMostOnce();
+        offHeapBuffer = null; // to enable GC of unsafe memory
+    }
+
+    private synchronized void runCleanerAtMostOnce() {
         if (cleaner != null) {
             cleaner.run();
         }
-        offHeapBuffer = null; // to enable GC of unsafe memory
         cleaner = null;
     }
```

Comment:
I understand the fix. But aren't we solving the problem at the wrong place? As far as I understand the `HybridMemorySegment` is not thread safe and should only be used by a single thread because otherwise we end up with corrupted data. Given this, why do we allow concurrent/multiple freeing operations to happen in the first place?